### PR TITLE
[OpRefactor] Update the generator to raise an exception if not defined

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Recent PRs have reintroduced exception handling for operator methods.

**Description of the Change:**

- The `generator` method has been moved from `Operation` to `Operator`, as I noticed there are a lot of places in the codebase where `op.generator()` is called, and sometimes it is unknown whether `op` is an `Operation` or `Operator`.

- `op.generator()` returns `GeneratorUndefinedError` if no generator is defined.

**Benefits:**

- It is now easy to chain operator methods (`op.generator().matrix()`) without worrying about returned types.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -323,6 +323,16 @@ def classproperty(func):
     return ClassPropertyDescriptor(func)
 
 
+# =============================================================================
+# Error classes
+# =============================================================================
+
+
+class GeneratorUndefinedError(Exception):
```
Comment (mariaschuld) [inline pennylane/operation.py:331]: I like this - we can use this pattern for all errors...

`DecompositionUndefinedError`
`MatrixUndefinedError`
`SparseMatrixUndefinedError`
`EigvalsUndefinedError`
...

File: pennylane/utils.py [hunk 2]
```diff
@@ -436,10 +436,7 @@ def get_generator(op, return_matrix=False):
     This utility function should be removed once the aforementioned classes
     no longer differ in behaviour.
     """
-    gen = getattr(op, "generator", lambda: None)()
-
-    if gen is None:
-        raise ValueError(f"Operation {op.name} does not have a generator")
+    gen = op.generator()
```
Comment (mariaschuld) [inline pennylane/utils.py:439]: This is nicer!

**Review by mariaschuld — APPROVED** [review]
Awesome!

File: pennylane/utils.py (hunk 2 above)
Comment (josh146) [reply pennylane/utils.py:439]: yep I am a fan! The generator being undefined is now treated at the source, and it is totally up to the developer if they want to catch it, or leave it and let the error propagate.

Comment (josh146) [reply pennylane/utils.py:439]: Another nice artifact: it avoids the issue where _slightly different_ error messages are defined in differnent places in the code base. 

For example, in `master`, you have `Operation does not have a generator` in the metric tensor code, and `Generator not defined for Operation` in the Fourier module 😆

**Pull request merged**